Add PvD over DNS support

NTag opened the pull request:
Here is my proposition to add PvD over DNS support to NEAT, according to [draft-stenberg-mif-mpvd-dns-00](https://tools.ietf.org/html/draft-stenberg-mif-mpvd-dns-00).

I had to store the `prefix_length` of an address, and PvD information in `struct neat_addr`.

When a new address becomes available, `neat_pvd.c:neat_pvd_handle_newaddr(...)` is called. It sends a PTR request to every DNS server, and for each answer sends a request for TXT records matching the answer. As specified in [draft-stenberg-mif-mpvd-dns-00](https://tools.ietf.org/html/draft-stenberg-mif-mpvd-dns-00#section-3.2.1) each TXT record is a piece of information (format is `key=value`) which is added to a `struct pvd_infos` which at the end is added to the list of PvDs attached to an address (`neat_addr.pvds`).

It is my first contribution to the project, so don't hesitate to make comments.


Here is an example of output in debug mode:
```
[INF] Available src-addresses:
[INF] 	IPv6: fd17:625d:a138:2:a00:27ff::86b6/64 pref 4294967295 valid 4294967295
[INF] 	IPv6: 2001:db8:bad:a55::198/64 pref 0 valid 4294967295
[INF] 		PVD:
[INF] 			r => 2001:db8:bad:a55::2
[INF] 			z => neat.eu
[INF] 		PVD:
[INF] 			z => cisco.com,mail.cisco.com
[INF] 			6 => 2001:db8:bad:a55::/64
[INF] 	IPv4: 192.168.55.198/24
[INF] 		PVD:
[INF] 			r => 2001:db8:bad:a55::2
[INF] 			z => neat.eu
[INF] 		PVD:
[INF] 			s => (null)
[INF] 	IPv4: 10.0.2.15/24
```

kristrev commented:
Great! I will take a look at this code tomorrow. I will not be able to test it though, but I guess you have done that ;)

tuexen commented:
Thanks. We can't use the buildbots for checking right now, since they are still red due to HE issues. I hope that this get cleared up soon.

NTag commented:
@kristrev Yes, I build and run the examples, I did not encounter any problems, but of course this did not cover every case.

kristrev commented:
That should not be an issue. I will focus on the integration part then.

On Thursday, 19 May 2016, Basile Bruneau <notifications@github.com> wrote:

> @kristrev <https://github.com/kristrev> Yes, I build and run the
> examples, I did not encounter any problems, but of course this did not
> cover every case.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/NEAT-project/neat/pull/80#issuecomment-220343668>
>

kristrev commented on neat_linux.c:
```diff
@@ -102,7 +102,7 @@ static void neat_linux_handle_addr(struct neat_ctx *nc,
     //addresses handlers/types of context?
     neat_addr_update_src_list(nc, &src_addr, ifm->ifa_index,
-            nl_hdr->nlmsg_type == RTM_NEWADDR, ifa_pref, ifa_valid);
+            nl_hdr->nlmsg_type == RTM_NEWADDR, ifm->ifa_prefixlen, ifa_pref, ifa_valid);
```
This change needs to be done in neat_bsd.c. If you don't have access to a BSD machine, then insert 0 as a temporary value and let @tuexen or @weinrank know. They can probably tell you which variable to read.

NTag replied:
I insert `0` at the moment with e8e0e9510041fb8092952547b6965c3bb623ebae, but it is temporary of course.

kristrev commented on neat_he.c:
```diff
@@ -215,6 +215,9 @@ neat_error_code neat_he_lookup(neat_ctx *ctx, neat_flow *flow, uv_poll_cb callba
         ctx->resolver = neat_resolver_init(ctx, "/etc/resolv.conf",
                                            he_resolve_cb, NULL);
     }
+    if (!ctx->pvd) {
```
I would suggest removing brackets here.

kristrev commented on neat_addr.c:
```diff
@@ -33,6 +33,15 @@ static void neat_addr_print_src_addrs(struct neat_ctx *nc)
         }
+
+        struct pvd* pvd;
+        struct pvd_info* pvd_info;
```
Please declare these at the top of the function.

kristrev commented on neat_pvd.c:
```diff
@@ -2,17 +2,199 @@
+    ldns_rr *rr;
+    ldns_rdf *dns_record = NULL;
+
+    struct pvd* pvd = (struct pvd *) malloc(sizeof(struct pvd));
```
Missing check for OOM.

kristrev commented on neat_addr.h:
```diff
@@ -38,6 +39,8 @@ struct neat_addr {
     uint16_t __pad2;
     uint8_t prefix_length;
+
+    struct pvds pvds;
```
I meant the opposite. That there is a struct in neat_pvd.h that stores a pointer to the address and the result of the PVD query. Not sure if it makes sense, but then again this code will be changed when the policy manager is up and running. Until the PM is ready, I would like to isolate this code as much as possible. So I suggest we have a struct pvd_res (or something) in neat_pvd.h that contains the PVD-list and a pointer to the address. If tying the PVD to an address is required, then a least make this a pointer to the pvd_res struct.

NTag replied:
Ok, I wasn't sure you meant that.

So I did what you propose in bbe3202, everything is stored in `neat_pvd->results`. And when we want to access the PvDs of an address, we simply iterate on the results until we find the one corresponding.

kristrev commented on neat_pvd.c:
```diff
@@ -2,17 +2,199 @@
+        ldns_resolver_push_nameserver(resolver, dns_src);
+
+        // Performing DNS query to 'dns_addr', for PTR records of 'reverse_ip'
+        p = ldns_resolver_query(resolver, domain, LDNS_RR_TYPE_PTR, LDNS_RR_CLASS_IN, LDNS_RD);
```
This is a blocking request. neat_resolver.c contains code showing how to do async DNS queries. Basically what you need to do is to create a handle + pointer to data structure for each request, store the pointer in the handle and then wait for the callback to be called and parse the result.

kristrev commented:
I had a look now and the logic itself looks fine. However, the code is not asynchronous, which is a requirement for the code to be merged into NEAT. Also, I am unsure if it will compile/work on BSD, so that will have to be tested as well. I also have some comments here and there.

weinrank commented:
The code only compiles on Linux - OSX, FreeBSD and NetBSD fail.

tuexen commented:
... as you can see from http://buildbot.nplab.de:28010/waterfall?branch=ntag/pvds

NTag commented:
Thank you very much for your valuable comments!

I still have to rewrite the code in an asynchronous way, and there are still some memory problems; but I think I addressed the other remarks.

NTag commented:
I did a few commits and tried to fulfill your comments. The code should now be asynchronous and compile under BSD.

As you can see in the build bots, there are just a few memory leaks left, which I am trying to correct.

kristrev commented:
I just went through the code. Since I am not able to test PvD itself (no supported equipment) I can't validate the functionality, but the code looks ok now (and is async). I see some style issues, but no point fixing them before we have the final coding style locked down.

evyncke commented:
Working code is good but working code with style is better ;-)

From: Kristian Evensen <notifications@github.com<mailto:notifications@github.com>>
Reply-To: NEAT-project/neat <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Wednesday 22 June 2016 at 18:15
To: NEAT-project/neat <neat@noreply.github.com<mailto:neat@noreply.github.com>>
Subject: Re: [NEAT-project/neat] Add PvD over DNS support (#80)


I just went through the code. Since I am not able to test PvD itself (no supported equipment) I can't validate the functionality, but the code looks ok now (and is async). I see some style issues, but no point fixing them before we have the final coding style locked down.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/NEAT-project/neat/pull/80#issuecomment-227795876>, or mute the thread<https://github.com/notifications/unsubscribe/AEW9H91P6u6BYKvbIAOdDWjIdi-NipZrks5qOV-PgaJpZM4IdIyi>.

NTag commented:
I am sorry @oystedal, your comments left with by rebase, but I addressed it with 8933be0 ;-) .

NTag commented:
Something changed with buildbots, isn't? Yesterday they were launching `./client_http_get bsd10.nplab.de /cgi-bin/he` and today they try `./client_http_get -u /cgi-bin/he bsd10.nplab.de` which fail with `usage: neat_http_get HOST [URI]`.

NTag commented:
Ok, finally buildbots are green :) !

NTag commented:
So, @kristrev @tuexen, can we merge it, or is there still something I have to do?

kristrev commented:
I have no objections.

kristrev commented:
Or, you might want to postpone this a bit. We are currently reworking the resolver, so it could be that there will be some minor changes in PvD. Hopefully, they should be minor.

kristrev commented:
Resolver rework is merged to master, so feel free to fix conflicts and merge. The most major changes are (from the top of my head).

* New signature for resolve callback.
* getaddrinfo() has been renamed (to neat_resolve) and changed slightly. For example, you pass a void pointer that is returned to the callback.

The rest should be more or less the same.

NTag commented:
Ok @kristrev, it's done (I just had to add a new inclusion). valgrind shows still no leaks, but maybe we should wait for the return of the buildbots to be sure everything is ok, and I think we will be able to merge it :) !

oystedal commented on neat_pvd.c:
```diff
@@ -0,0 +1,531 @@
+
+        struct sockaddr_storage *dns_addr = &(dns_server->server_addr);
+
+        struct pvd_dns_query *dns_query = malloc(sizeof(struct pvd_dns_query));
```
Missing a NULL check on this malloc

oystedal commented on neat_pvd.c:
```diff
@@ -0,0 +1,531 @@
+                      size_t suggested_size,
+                      uv_buf_t *buf)
+{
+    char *dns_rcv_buf = calloc(sizeof(char), 1472);
```
1472 seems like a magic number. Is it based on anything specific or is it an estimate?

NTag replied:
I changed to DNS_BUF_SIZE from the resolver.

oystedal commented on neat_pvd.c:
```diff
@@ -0,0 +1,531 @@
+        }
+    } else {
+        server_addr6                        = (struct sockaddr_in6 *) dns_addr;
+        async_query->dst_addr6              = calloc(sizeof(struct sockaddr_in6), 1);
```
Missing NULL check

oystedal commented on neat_pvd.c:
```diff
@@ -0,0 +1,531 @@
+
+    if (dns_addr2->sa_family == AF_INET) {
+        server_addr4                        = (struct sockaddr_in *) dns_addr;
+        async_query->dst_addr4              = calloc(sizeof(struct sockaddr_in), 1);
```
Missing NULL check

oystedal commented on neat_pvd.c:
```diff
@@ -0,0 +1,531 @@
+
+    async_query->dns_uv_snd_buf = calloc(sizeof(uv_buf_t), 1);
+    async_query->dns_snd_handle = calloc(sizeof(uv_udp_send_t), 1);
+    async_query->resolve_handle = calloc(sizeof(uv_udp_t), 1);
```
Missing NULL check

oystedal commented on neat_pvd.c:
```diff
@@ -0,0 +1,531 @@
+{
+    struct in_addr src_addr4;
+    struct in6_addr src_addr6;
+    char reverse_ip[80]; // maximum length for a reverse /128 IPv6
```
Is this sufficient to store the maximum length of the reverse address plus an additional 0-byte (which is added by sprintf) at the end?

NTag replied:
Yes I think, why wouldn't it be?

oystedal replied:
Just making sure it has been considered, is all

NTag replied:
Yes, you are right; and I have tested it.

oystedal commented:
Had another read through the code. Looks great, I could only find very minor things to be picky about :)

That being said, I do not know much about PvD, so I haven't looked much at the correctness of the implementation. I assume you have tested this.

Great work! :)

NTag commented:
Thank you @oystedal, I think I have addressed your comments.

I am almost sure I have correctly implemented the protocol (and I have tested it of course), but if it wasn't the case I would fix it ;-) .

NTag commented:
Fixed @oystedal. I think this time I have removed all remaining leaks :-) .

oystedal commented:
The buildbots seem happy now, so I think the code is ready to be merged. :)

NTag commented:
I am going to fix the new conflicts, verify the buildbots and merge it so :)
